### PR TITLE
OpenZFS 7361 Improve L2ARC filling

### PR DESCRIPTION
**Pull request by kernelOfTruth**
This patch modifies ZFS so that when it has to evict buffers from ARC, it first chooses buffers which are already in L2ARC.

Authored by: Ben RUBSON <ben.rubson@gmail.com>
Reviewed by:
Reviewed by:
Approved by:
Ported-by: kernelOfTruth kerneloftruth@gmail.com
OpenZFS-issue: https://www.illumos.org/issues/7361
OpenZFS-commit:

Porting notes:

account for "‘for’ loop initial declarations are only allowed in C99 or C11 mode":

int l, i;

### CODE REVIEW

**Comment by kernelOfTruth** [comment]
https://www.illumos.org/issues/7361

The improvements look pretty great, I wonder if the effect persist when e.g. transferring data from one pool to the other, exporting the 2nd pool, importing a 3rd one (while the 1st one is in usage) and then re-transferring the same (similar) data again via e.g. rsync

**Comment by benrubson** [comment]
@kernelOfTruth : if your ARC has to evict during the first transfer, because it is too small to retain all (meta)data, then it will only evict (meta)data which is already in L2ARC.
So the next data transfers will find all the (meta)data they need in the L2ARC, which will speed them up.

Some considerations :
- of course your L2ARC has to be big enough to retain all this (meta)data, on huge file servers perhaps `secondarycache=metadata` is good enough, and it will for sure speed-up things such as rsync ;
- L2ARC has to be fed fast enough to avoid ZFS throttle waiting for some ARC free space, so you may have to tune `l2arc_headroom` and `l2arc_feed_min_ms`.

**Comment by benrubson** [comment]
_Just as a reference here is the OpenZFS PR : https://github.com/openzfs/openzfs/pull/189_

**Comment by kernelOfTruth** [comment]
@benrubson thanks, that sound promising, indeed and might perfectly cover exact my kind of use case :)

File: module/zfs/arc.c [hunk 1]
```diff
@@ -3221,7 +3223,9 @@ arc_evict_state_impl(multilist_t *ml, int idx, arc_buf_hdr_t *marker,
 		ASSERT(!MUTEX_HELD(hash_lock));
 
 		if (mutex_tryenter(hash_lock)) {
-			uint64_t evicted = arc_evict_hdr(hdr, hash_lock);
+			uint64_t evicted = 0;
+			if (bytes == ARC_EVICT_ALL || marker->b_lsize == 2 || HDR_HAS_L2HDR(hdr) || (HDR_PREFETCH(hdr) && l2arc_noprefetch))
```
Comment (rlaager) [inline module/zfs/arc.c]: This needs some line wrapping.

Comment (benrubson) [inline module/zfs/arc.c]: Just done in the original PR https://github.com/openzfs/openzfs/pull/189

**Comment by behlendorf** [comment]
Closing until the patch is finalized, discussion on going in openzfs/openzfs#189

**Pull request closed without merging**